Neos.Flow/Classes/Mvc/Routing/Routes.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+namespace Neos\Flow\Mvc\Routing;
+
+use Neos\Flow\Annotations as Flow;
+use Neos\Flow\Configuration\ConfigurationManager;
+use Neos\Flow\Mvc\Exception\InvalidRouteSetupException;
+use Traversable;
+
+/**
+ * @Flow\Proxy(false)
+ */
+final class Routes implements \IteratorAggregate
+{
+    /**
+     * @var array<int, Route>
+     */
+    private readonly array $routes;
+
+    private function __construct(
+        Route ...$routes
+    ) {
+        $this->routes = $routes;
+    }
+
+    public static function fromConfiguration(array $configuration): self
+    {
+        $routes = self::empty();
+        $routesWithHttpMethodConstraints = [];
+        foreach ($configuration as $routeConfiguration) {
+            $route = Route::fromConfiguration($routeConfiguration);
+
+            $uriPattern = $route->getUriPattern();
+
+            if (isset($routeConfiguration['httpMethods'])) {
+                if (isset($routesWithHttpMethodConstraints[$uriPattern]) && $routesWithHttpMethodConstraints[$uriPattern] === false) {
+                    throw new InvalidRouteSetupException(sprintf('There are multiple routes with the uriPattern "%s" and "httpMethods" option set. Please specify accepted HTTP methods for all of these, or adjust the uriPattern', $uriPattern), 1365678427);
```

Comment:
imo validation belongs to the actual constructor, so that these constrains are also validated for `prepend` `append` and any other future methods ;)

Reply:
@mhsdesign So adding a `validate` method after the assigning of `$this->routes` in the private constructor?

Reply:
> imo validation belongs to the actual constructor

Oh yes, that's a very good point!

I'd say something along the lines of:

```php
    private function __construct(
        Route ...$routes
    ) {
        $routesWithHttpMethodConstraints = [];
        foreach ($this->routes as $route) {
            $uriPattern = $route->getUriPattern();
            if ($route->hasHttpMethodConstraints()) {
                if ($routesWithHttpMethodConstraints[$uriPattern] ?? null === false) {
                    throw new InvalidRouteSetupException(sprintf('There are multiple routes with the uriPattern "%s" and "httpMethods" option set. Please specify accepted HTTP methods for all of these, or adjust the uriPattern', $uriPattern), 1365678427);
                }
                $routesWithHttpMethodConstraints[$uriPattern] = true;
        } else {
            if ($routesWithHttpMethodConstraints[$uriPattern] ?? null === true) {
                throw new InvalidRouteSetupException(sprintf('There are multiple routes with the uriPattern "%s" and "httpMethods" option set. Please specify accepted HTTP methods for all of these, or adjust the uriPattern', $uriPattern), 1365678432);
            }
            $routesWithHttpMethodConstraints[$uriPattern] = false;
        }
    }
```

And then change this method to just:
```php
    public static function fromConfiguration(array $configuration): self
    {
        $routes = [];
        foreach ($configuration as $routeConfiguration) {
            $routes[] = Route::fromConfiguration($routeConfiguration);
        }
        return new self(...$routes);
    }
```

(untested)

Reply:
Exactly what i had in mind ❤️

Reply:
Adjusted ;)